src/ui/public/courier/data_source/_abstract.js

```diff
@@ -287,20 +287,26 @@ export function AbstractDataSourceProvider(Private, Promise, PromiseEmitter) {
           };
         }
 
-        if (flatState.body.size > 0) {
-          const computedFields = flatState.index.getComputedFields();
-          flatState.body.stored_fields = computedFields.storedFields;
-          flatState.body.script_fields = flatState.body.script_fields || {};
-          flatState.body.docvalue_fields = flatState.body.docvalue_fields || [];
-
-          _.extend(flatState.body.script_fields, computedFields.scriptFields);
-          flatState.body.docvalue_fields = _.union(flatState.body.docvalue_fields, computedFields.docvalueFields);
-
-          if (flatState.body._source) {
-            // exclude source fields for this index pattern specified by the user
-            const filter = fieldWildcardFilter(flatState.body._source.excludes);
-            flatState.body.docvalue_fields = flatState.body.docvalue_fields.filter(filter);
-          }
+        const computedFields = flatState.index.getComputedFields();
+        flatState.body.stored_fields = computedFields.storedFields;
+        flatState.body.script_fields = flatState.body.script_fields || {};
+        flatState.body.docvalue_fields = flatState.body.docvalue_fields || [];
+
+        _.extend(flatState.body.script_fields, computedFields.scriptFields);
+        flatState.body.docvalue_fields = _.union(flatState.body.docvalue_fields, computedFields.docvalueFields);
+
+        if (flatState.body._source) {
+          // exclude source fields for this index pattern specified by the user
+          const filter = fieldWildcardFilter(flatState.body._source.excludes);
+          flatState.body.docvalue_fields = flatState.body.docvalue_fields.filter(filter);
+        }
+
+        // if we only want to search for certain fields
+        const fields = flatState.fields;
+        if (fields) {
+          flatState.body.docvalue_fields = _.intersection(flatState.body.docvalue_fields, fields);
+          flatState.body.script_fields = _.pick(flatState.body.script_fields, fields);
+          flatState.body._source = _.difference(fields, [...flatState.body.docvalue_fields, ..._.keys(flatState.body.script_fields)]);
```

Comment:
Thanks @Bargs, you've really helped out a ton with these changes!
> I wouldn't remove the docvalue_fields from _source. We don't know for sure that these fields have doc_values, it's more of a supplement to _source
Great point, will do.
>I think we should probably leave _source.excludes intact if it exists. Instead of assigning directly to _source you could assign to _source.includes
Another great point, I will make it so.
>Why haven't you set flatState.body.stored_fields as well?
This is currently being hard-coded to `['*']` [here](https://github.com/elastic/kibana/blob/9f33293e5c45f09feecfefd0c0a0ea9e14317d2f/src/ui/public/index_patterns/_get_computed_fields.js#L21). I could potentially include the same fields that we're requesting from `_source` here, do you think that'd help performance?